[V1.6.0 -> V1.7.4] Error reading configuration file, java.util.zip.ZipException: invalid code lengths set.

## Expected Behavior
Detekt gradle task finish successfully when there are changes to custom `RuleProvider` or `Rule`.

## Observed Behavior
Detekt gralde task failed with the following stack. 

```
Caused by: java.util.ServiceConfigurationError: io.gitlab.arturbosch.detekt.api.RuleSetProvider: Error reading configuration file
        at io.gitlab.arturbosch.detekt.core.RuleSetLocator.load(RuleSetLocator.kt:29)
        at io.gitlab.arturbosch.detekt.cli.config.DefaultPropertiesConfigValidator$validate$1.invoke(DefaultPropertiesConfigValidator.kt:20)
        at io.gitlab.arturbosch.detekt.cli.config.DefaultPropertiesConfigValidator.validate(DefaultPropertiesConfigValidator.kt:27)
        at io.gitlab.arturbosch.detekt.cli.config.ConfigValidatorsKt.checkConfiguration(ConfigValidators.kt:15)
        at io.gitlab.arturbosch.detekt.cli.runners.Runner.execute(Runner.kt:31)
        at io.gitlab.arturbosch.detekt.invoke.DefaultCliInvoker.invokeCli(DetektInvoker.kt:51)
        ... 89 more
Caused by: java.util.zip.ZipException: invalid code lengths set
        ... 95 more
```


## Steps to Reproduce
1) Remove a rule from a `RuleProvider`
2) Run gradle detekt task.

## Your Environment
* Version of detekt used: v1.7.4
* Version of Gradle used (if applicable): v5.6.4
* Operating System and version: Mac OSX 10.14 Mojave and Arch 5.5.4-arch1-1

--- a/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/ClassLoaderCache.kt
+++ b/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/ClassLoaderCache.kt
@@ -14,14 +14,21 @@ object ClassLoaderCache {
 
         if (lastClasspath == null) {
             cache(classpath)
-        } else {
-            if (!lastClasspath.minus(classpath).isEmpty) {
-                DefaultGroovyMethodsSupport.closeQuietly(lastLoader)
-                cache(classpath)
-            }
+        } else if (hasClasspathChanged(lastClasspath, classpath)) {
+            DefaultGroovyMethodsSupport.closeQuietly(lastLoader)
+            cache(classpath)
         }
 
         return loaderAndClasspath?.first ?: error("Cached or newly created detekt classloader expected.")
+    }
+
+    internal fun hasClasspathChanged(lastClasspath: FileCollection, currentClasspath: FileCollection): Boolean {
+        if (lastClasspath.files.size != currentClasspath.files.size) {
+            return true
+        }
+        return lastClasspath.sorted()
+            .zip(currentClasspath.sorted())
+            .any { (last, current) -> last != current || last.lastModified() != current.lastModified() }
     }
 
     private fun cache(classpath: FileCollection) {

--- a/detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/internal/ClassLoaderCacheSpec.kt
+++ b/detekt-gradle-plugin/src/test/kotlin/io/gitlab/arturbosch/detekt/internal/ClassLoaderCacheSpec.kt
@@ -1,0 +1,82 @@
+package io.gitlab.arturbosch.detekt.internal
+
+import org.assertj.core.api.Assertions.assertThat
+import org.gradle.api.internal.file.AbstractFileCollection
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+import java.io.File
+import kotlin.random.Random
+
+internal class ClassLoaderCacheSpec : Spek({
+
+    describe("classpath changes") {
+
+        it("passes for same files") {
+            val changed = ClassLoaderCache.hasClasspathChanged(
+                TestFileCollection(FixedDateFile("/a/b/c")),
+                TestFileCollection(FixedDateFile("/a/b/c"))
+            )
+
+            assertThat(changed).isFalse()
+        }
+
+        it("reports for different file count") {
+            val changed = ClassLoaderCache.hasClasspathChanged(
+                TestFileCollection(DifferentDateFile("/a/b/c"), DifferentDateFile("/c/b/a")),
+                TestFileCollection(DifferentDateFile("/a/b/c"))
+            )
+
+            assertThat(changed).isTrue()
+        }
+
+        it("reports different files") {
+            val changed = ClassLoaderCache.hasClasspathChanged(
+                TestFileCollection(DifferentDateFile("/c/b/a")),
+                TestFileCollection(DifferentDateFile("/a/b/c"))
+            )
+
+            assertThat(changed).isTrue()
+        }
+
+        it("reports same files with different modify date") {
+            val changed = ClassLoaderCache.hasClasspathChanged(
+                TestFileCollection(DifferentDateFile("/a/b/c")),
+                TestFileCollection(DifferentDateFile("/a/b/c"))
+            )
+
+            assertThat(changed).isTrue()
+        }
+    }
+})
+
+private class TestFileCollection(private vararg val files: File) : AbstractFileCollection() {
+
+    override fun getFiles(): MutableSet<File> = files.toMutableSet()
+    override fun getDisplayName(): String = "FileCol"
+}
+
+private open class FixedDateFile(path: String) : File(path) {
+    override fun compareTo(other: File?): Int {
+        if (other == null) return 1
+        return path.compareTo(other.path)
+    }
+
+    override fun lastModified(): Long = 12345L
+}
+
+private class DifferentDateFile(path: String) : FixedDateFile(path) {
+
+    companion object {
+        private val random = Random(seed = 200)
+        private val cache = HashSet<Long>()
+    }
+
+    override fun lastModified(): Long {
+        var nextDate = random.nextLong()
+        while (cache.contains(nextDate)) {
+            nextDate = random.nextLong()
+        }
+        cache.add(nextDate)
+        return nextDate
+    }
+}

